File: Python/Array/Union.py
from collections import defaultdict
class Solution:
    # Better approach 2: O((m+n)log(m+n)), O(m+n) [Using map/dict]
    # def union_array(self,arr1,arr2):
    #     
    #     freq=defaultdict(int)
    #     for i in arr1:
    #         freq[i]+=1
        
    #     for j in arr2:
    #         freq[j]+=1
        
    #     ans=list()
    #     for x,y in freq.items():
    #         ans.append(x)
        
    #     return ans
    
    #  Better approach 2: O((m+n)),O(m+n) [Using set, cause in python set is hash table which insertion take O(1)]
    # def union_array(self,arr1,arr2):
    #     temp=set()
    #     ans=list()
    #     for i in arr1:
    #         temp.add(i)

    #     for j in arr2:
    #         temp.add(j)

    #     for x in temp:
    #         ans.append(x)
        
    #     return ans
    
    # optimal approach: O(n+m),O(n+m) 
    def union_array(self,arr1,arr2):
        n=len(arr1)
        m=len(arr2)
        ans=list()
        i,j=0,0
        while i<n and j<m:
            if arr1[i]==arr2[j]:
                if not ans or ans[-1]!=arr1[i]:
                    ans.append(arr1[i])
                i+=1
                j+=1

            elif arr1[i]<arr2[j]:
                if not ans or ans[-1]!=arr1[i]:
                    ans.append(arr1[i])
                i+=1

            elif arr2[j]<arr1[i]:
                if not ans or ans[-1]!=arr2[j]:
                    ans.append(arr2[j])
                j+=1

        while i<n:
            if not ans or ans[-1]!=arr1[i]:
                    ans.append(arr1[i])
            i+=1
            
        while j<m:
            if not ans or ans[-1]!=arr2[j]:
                    ans.append(arr2[j])
            j+=1

        return ans


if __name__=="__main__":
    arr1=[1,2,3,4,5,6]
    arr2=[2,4,6,8,9]

    sol=Solution()
    ans=sol.union_array(arr1,arr2)
    print(ans)

File: CPP/Array/Union.cpp
/*
 * DATE: April 5, 2026
 * PATTERN: Hashing, Two Pointers, Sorting
 *
 * APPROACHES:
 * 1. BRUTE FORCE:
 * - Step 1: Concatenate both arrays into a list and remove duplicates by checking each element against all seen (nested loops).
 * - Step 2: Sort the result if order is required.
 * - TIME: O((n+m)^2) worst naive dedup | SPACE: O(n+m)
 *
 * 2. BETTER:
 * - Step 1: Insert all elements from both arrays into a set (or map keys) for uniqueness.
 * - Step 2: Copy sorted order from set/map into answer vector.
 * - TIME: O((n+m) log(n+m)) | SPACE: O(n+m) for the set
 *
 * 3. OPTIMAL (when both arrays are already sorted):
 * - Step 1: Merge-like two pointers, skipping duplicates within each array and across pointers.
 * - Step 2: Push smaller head, advance; on equality advance both.
 * - TIME: O(n+m) | SPACE: O(1) extra besides output
 *
 * AHA-MOMENT: Uniqueness is a set problem; sorted inputs unlock linear two-pointer merge without global sort.
 */

#include<bits/stdc++.h>
using namespace std;

class Solution
{
    public:
    // Better approach 2: O((m+n)log(m+n)), O(m+n) [Using map]
    vector<int> union_array(vector<int>& arr1, vector<int> &arr2, int n, int m)
    {
        map<int,int> mp;
        vector<int> ans;
        
        for(int i=0;i<n;i++)
        {
            mp[arr1[i]]++;
        }

        for(int i=0;i<m;i++)
        {
            mp[arr2[i]]++;
        }

        for(auto x:mp)
        {
            ans.push_back(x.first);
        }
        return ans;
    }

    

    // Better approach 2: O((m+n)log(m+n)),O(m+n) [Using set]
    vector<int> union_array(vector<int>& arr1, vector<int> &arr2, int n, int m)
    {
        set<int> st;
        vector<int>ans;
        for(int i=0;i<n;i++)
        {
            st.insert(arr1[i]);
        }

        for(int i=0;i<m;i++)
        {
            st.insert(arr2[i]);
        }

        for(int x:st)
        {
            ans.push_back(x);
        }
        return ans;
        
    }

    // Optimal Solution: O(n+m),O(n+m)
    vector<int> union_array(vector<int>& arr1, vector<int> &arr2, int n, int m)
    {
        int i=0;
        int j=0;
        vector<int> ans;
        while(i<n && j<m)
        {
            if(arr1[i]==arr2[j])
            {
                if(ans.empty()|| ans.back()!=arr1[i])
                {
                    ans.push_back(arr1[i]);
                }
                i++;
                j++;
            }

            else if(arr1[i]<arr2[j])
            {
                if(ans.empty()|| ans.back()!=arr1[i])
                {
                    ans.push_back(arr1[i]);
                }
                i++;
                
            }
            else{
                 if(ans.empty()|| ans.back()!=arr2[j]){
                        ans.push_back(arr2[j]);
                 }
                 j++;
            }
        }
        while(i<n)
        {
             if(ans.empty()|| ans.back()!=arr1[i])
            {
                ans.push_back(arr1[i]);
            }
            i++;
        }

        while(j<m)
        {
             if(ans.empty()|| ans.back()!=arr2[j])
            ans.push_back(arr2[j]);
            j++;
        }

        return ans;
    }
};

int main()
{
    Solution sol;
    vector<int> arr1={1,2,3,4,5};
    vector<int> arr2={2,3,4,4,5};
    vector<int> ans;
    ans=sol.union_array(arr1,arr2,arr1.size(),arr2.size());
    for(int x:ans)
    {
        cout<<x<<" ";
    }
    return 0;
}